code/controllers/subsystem/dbcore.dm

```diff
@@ -272,131 +235,78 @@ Delayed insert mode was removed in mysql 7 and only works with MyISAM type table
+	var/list/item
+	var/last_error
 
-/datum/DBQuery/New(sql_query, datum/BSQL_Connection/connection)
+/datum/DBQuery/New(sql_query, sql_parameters)
 	SSdbcore.active_queries[src] = TRUE
-	Activity("Created")
-	item = list()
-	src.connection = connection
 	sql = sql_query
+	parameters = sql_parameters
 
 /datum/DBQuery/Destroy()
-	Close()
 	SSdbcore.active_queries -= src
 	return ..()
 
 /datum/DBQuery/CanProcCall(proc_name)
 	//fuck off kevinz
 	return FALSE
 
-/datum/DBQuery/proc/SetQuery(new_sql)
+/datum/DBQuery/proc/SetQuery(new_sql, new_parameters)
 	if(in_progress)
 		CRASH("Attempted to set new sql while waiting on active query")
-	Close()
 	sql = new_sql
-
-/datum/DBQuery/proc/Activity(activity)
-	last_activity = activity
-	last_activity_time = world.time
+	parameters = new_parameters
 
 /datum/DBQuery/proc/warn_execute(async = TRUE)
 	. = Execute(async)
 	if(!.)
 		to_chat(usr, "<span class='danger'>A SQL error occurred during this operation, check the server logs.</span>")
 
 /datum/DBQuery/proc/Execute(async = TRUE, log_error = TRUE)
-	Activity("Execute")
 	if(in_progress)
-		CRASH("Attempted to start a new query while waiting on the old one")
+		CRASH("Attempted to start a new query while waiting on the old one.")
 
-	if(QDELETED(connection))
-		last_error = "No connection!"
-		return FALSE
-
-	var/start_time
-	var/timed_out
-	if(!async)
-		start_time = REALTIMEOFDAY
-	Close()
-	timed_out = run_query(async)
-	if(query.GetErrorCode() == 2006) //2006 is the return code for "MySQL server has gone away" time-out error, meaning the connection has been lost to the server (if it's still alive)
-		log_sql("Executing query encountered returned a lost database connection (2006).")
-		SSdbcore.Disconnect()
-		if(SSdbcore.Connect()) //connection was restablished, reattempt the query
-			log_sql("Connection restablished")
-			timed_out = run_query(async)
-		else
-			log_sql("Executing query failed to restablish database connection.")
-	skip_next_is_complete = TRUE
-	var/error = QDELETED(query) ? "Query object deleted!" : query.GetError()
-	last_error = error
-	. = !error
-	if(!. && log_error)
-		log_sql("[error] | Query used: [sql]")
-	if(!async && timed_out)
-		log_query_debug("Query execution started at [start_time]")
-		log_query_debug("Query execution ended at [REALTIMEOFDAY]")
-		log_query_debug("Slow query timeout detected.")
-		log_query_debug("Query used: [sql]")
-		slow_query_check()
+	log_sql("Executing query [sql] with parameters [parameters] at [REALTIMEOFDAY]. Is async: [async]")
+	run_query(async)
+	log_sql("Finished executing query [sql] with parameters [parameters] at [REALTIMEOFDAY]. Is async: [async]")
 
 /datum/DBQuery/proc/run_query(async)
-	query = connection.BeginQuery(sql)
-	if(!async)
-		. = !query.WaitForCompletion()
+	world.log << sql
+	world.log << parameters
+	in_progress = TRUE
+	if(async)
+		query_results = rustg_sql_query_async(sql, parameters)
+		world.log << query_results
+		var/done = 0
+		while(!done)

```

Comment:
you know you're supposed to sleep in this loop right?

Reply:
we're allowed to use sleep()?